### PR TITLE
Make Alphabet concept require noexcept

### PR DESCRIPTION
**Pull request by h-2**
@smehringer Can you give this a general review?

@rrahn I have two failing tests in pairwise assignment that are super strange. Everything else works... It is somehow related to my changes, but I don't see how. Can you have a look? Using the previous implementation if SEQAN3_IS_CONSTEXPR makes even more things fail – I also don't know why, but the new macro should really be more robust and even the weird composition tests all pass – just your alignment tests don't.

It would be great if we can resolve this quickly so I can do the actual changes of adding `assign_char_strict()`.

### CODE REVIEW

**Comment by h-2** [comment]
@rrahn it might be related to the gap alphabet still being an aggrgegate type and not yet refurbished in the new design. I'll check that out tomorrow.

File: include/seqan3/core/metafunction/function.hpp [hunk 1]
```diff
@@ -41,30 +41,24 @@
...
  * \returns true or false.
  */
-#if defined(__clang__) // bug: https://bugs.llvm.org//show_bug.cgi?id=15481
-#define SEQAN3_IS_CONSTEXPR(...) __builtin_constant_p( (__VA_ARGS__,0) )
-#else
-#define SEQAN3_IS_CONSTEXPR(...) noexcept(seqan3::detail::is_constexpr_helper( (__VA_ARGS__, 0) ))
-#endif
-
-} // namespace seqan3
+#define SEQAN3_IS_CONSTEXPR(...) seqan3::detail::is_constexpr_helper<(bool)__builtin_constant_p( (__VA_ARGS__,0) )>()
```
Comment (rrahn) [inline include/seqan3/core/metafunction/function.hpp]: do you still require the `,0`?

**Comment by rrahn** [comment]
@h-2 I'll check your branch.

**Comment by rrahn** [comment]
WTF 😱 

When I do:
```cpp
using gapped_chr = gapped<seq1_value_type>;
... std::vector<gapped_chr>...
```
then it works.

**Comment by rrahn** [comment]
Even if I leave the code as is and just introduce this at the beginning of selector.

```cpp
using gapped_dna4 [[maybe_unused]] = gapped<seq1_value_type>;
```
it works.

**Comment by rrahn** [comment]
@h-2  I made a PR against your branch with the fix for the compiler error.

File: include/seqan3/core/metafunction/function.hpp (hunk 1 above)
Comment (h-2) [reply include/seqan3/core/metafunction/function.hpp]: Yes, that's still required, because __VA_ARGS__ may resolve to void which results in an illegal expression to test.

**Comment by h-2** [comment]
> @h-2 I made a PR against your branch with the fix for the compiler error.

Wow, cool. I have no clue why, but it's good enough for now!

**Comment by h-2** [comment]
I think everything should work now. I have also simplified the macro further, getting rid of the helper function. 

@smehringer Can you give it a review?
@rrahn Can you merge if you agree?

Thanks!

File: include/seqan3/alphabet/concept.hpp [hunk 2]
```diff
@@ -98,17 +98,19 @@ concept semi_alphabet_concept = std::Regular<std::remove_reference_t<t>> &&
                                 std::StrictTotallyOrdered<t> &&
                                 requires (t t1, t t2)
 {
-
     // static data members
     alphabet_size<std::remove_reference_t<t>>::value;
     alphabet_size_v<std::remove_reference_t<t>>;
 
     // conversion to rank
-    { to_rank(t1) } -> underlying_rank_t<std::remove_reference_t<t>>;
```
Comment (smehringer) [inline include/seqan3/alphabet/concept.hpp]: why did you remove this in favour of the `std::Same`?

File: include/seqan3/alphabet/concept.hpp [hunk 3]
```diff
@@ -149,11 +151,20 @@ template <typename t>
...
     // assignment from char
-    { assign_char(t1,  0) }                          -> std::remove_reference_t<t> &;
-    { assign_char(std::remove_reference_t<t>{}, 0) } -> std::remove_reference_t<t> &&;
+    requires noexcept(assign_char(t1,  0));
+    requires std::Same<decltype(assign_char(t1,  0)), std::remove_reference_t<t> &>;
+    requires noexcept(assign_char(std::remove_reference_t<t>{},  0));
+    requires std::Same<decltype(assign_char(std::remove_reference_t<t>{}, 0)), std::remove_reference_t<t> &&>;
+
+    // assignment from char (strict)
+//     { assign_char_strict(t1,  0) } -> noexcept;
```
Comment (smehringer) [inline include/seqan3/alphabet/concept.hpp]: maybe add this when you add assign_char_strict but as out-commented code with no remark.

File: include/seqan3/alphabet/composition/cartesian_composition.hpp [hunk 4]
```diff
@@ -291,7 +291,7 @@ class cartesian_composition :
     /*!\name Constructors, destructor and assignment
      * \{
      */
-    constexpr cartesian_composition() : base_t{} {}
+    constexpr cartesian_composition() noexcept : base_t{} {}
     constexpr cartesian_composition(cartesian_composition const &) = default;
```
Comment (smehringer) [inline include/seqan3/alphabet/composition/cartesian_composition.hpp]: maybe dump question: why no `noexcept` here?

File: include/seqan3/alphabet/detail/member_exposure.hpp [hunk 5]
```diff
@@ -94,8 +94,8 @@ struct alphabet_size<alphabet_type_with_members>
  * \returns The letter's value in the alphabet's rank type (usually a `uint*_t`).
  */
 template <typename alphabet_type>
-constexpr underlying_rank_t<alphabet_type> to_rank(alphabet_type const alph)
-    requires requires (alphabet_type alph) { { alph.to_rank() } -> underlying_rank_t<alphabet_type>; }
+constexpr underlying_rank_t<alphabet_type> to_rank(alphabet_type const alph) noexcept
+    requires requires (alphabet_type alph) { { alph.to_rank() } -> underlying_rank_t<alphabet_type>;  }
```
Comment (smehringer) [inline include/seqan3/alphabet/detail/member_exposure.hpp]: remove extra space again :nail_care:

File: include/seqan3/alphabet/structure/structured_rna.hpp [hunk 6]
```diff
@@ -97,7 +97,7 @@ class structured_rna :
     /*!\name Constructors, destructor and assignment
      * \{
      */
-    constexpr structured_rna() : base_type{} {}
+    constexpr structured_rna() noexcept : base_type{} {}
     constexpr structured_rna(structured_rna const &) = default;
     constexpr structured_rna(structured_rna &&) = default;
     constexpr structured_rna & operator =(structured_rna const &) = default;
```
Comment (smehringer) [inline include/seqan3/alphabet/structure/structured_rna.hpp]: I cannot mark the proper line as it is not changed but you maybe want to add the noexcept specifier to the doxygen documentation in line 110? (copied docs from cartesian composition)

File: include/seqan3/alphabet/structure/structured_rna.hpp [hunk 7]
```diff
@@ -128,7 +128,7 @@ class structured_rna :
     //!\name Write functions
     //!\{
     //!\brief Assign from a nucleotide character. This modifies the internal sequence letter.
-    constexpr structured_rna & assign_char(char_type const c)
+    constexpr structured_rna & assign_char(char_type const c) noexcept
```
Comment (smehringer) [inline include/seqan3/alphabet/structure/structured_rna.hpp]: Should this be part of the PR as it does not consider the constructor? I checked that e.g. masked and quality do not have  the noexpect specifier at their assign_char function either..

File: include/seqan3/alphabet/concept.hpp (hunk 2 above)
Comment (smehringer) [reply include/seqan3/alphabet/concept.hpp]: ah ok. This might be an issue for several other concepts too...

File: include/seqan3/alphabet/composition/union_composition.hpp [hunk 8]
```diff
@@ -241,7 +241,7 @@ class union_composition : public alphabet_base<union_composition<alternative_typ
     /*!\name Constructors, destructor and assignment
      * \{
      */
-    constexpr union_composition() = default;
+    constexpr union_composition() noexcept = default;
```
Comment (smehringer) [inline include/seqan3/alphabet/composition/union_composition.hpp]: If this is defaulted is this also automatically noexcept? (regarding the comment before?)

Comment (h-2) [reply include/seqan3/alphabet/composition/union_composition.hpp]: Yes it would be, but it shouldn't be defaulted :scream: It should call `: base_type{} {}` !

Thanks for spotting this.

Comment (h-2) [reply include/seqan3/alphabet/composition/union_composition.hpp]: The wrappers did not check for noexcept on the member functions. I have added static assertions (they are more verbose than concept checks and it is something you can easily forget).

File: include/seqan3/alphabet/structure/structured_rna.hpp (hunk 7 above)
Comment (h-2) [reply include/seqan3/alphabet/structure/structured_rna.hpp]: fixed masked but didnt find an issue with quality...?

File: include/seqan3/alphabet/detail/member_exposure.hpp [hunk 9]
```diff
@@ -94,9 +94,15 @@ struct alphabet_size<alphabet_type_with_members>
  * \returns The letter's value in the alphabet's rank type (usually a `uint*_t`).
  */
 template <typename alphabet_type>
-constexpr underlying_rank_t<alphabet_type> to_rank(alphabet_type const alph)
+//!\cond
     requires requires (alphabet_type alph) { { alph.to_rank() } -> underlying_rank_t<alphabet_type>; }
+//!\endcond
+constexpr underlying_rank_t<alphabet_type> to_rank(alphabet_type const alph) noexcept
+
```
Comment (smehringer) [inline include/seqan3/alphabet/detail/member_exposure.hpp]: remove newline :nail_care: 
same in assign_rank, assign_char and complement below

File: include/seqan3/alphabet/structure/structured_rna.hpp (hunk 7 above)
Comment (smehringer) [reply include/seqan3/alphabet/structure/structured_rna.hpp]: sorry, it was qualified but you already fixed that.
